### PR TITLE
Restart openshift-apiserver only if aggregator-client-ca is expired

### PR DESCRIPTION
**Pull request by guillaumerose**
This cert is eventually renewed at startup. If it is renewed, we need to
wait for it and restart openshift-apiserver. Since this cert is located
both in a configmap and on the disk, we prefer to read it from the disk:
it is faster to access and more reliable.

If the cert is not expired, openshift-apiserver will correctly start.
`crc start` will take less time when the bundle is fresh (< 30 days).

follow up of https://github.com/code-ready/crc/pull/1532

### CODE REVIEW

File: pkg/crc/cluster/cluster.go [hunk 1]
```diff
@@ -67,6 +67,28 @@ func getcertExpiryDateFromVM(sshRunner *ssh.Runner) (time.Time, error) {
...
+	certExpiryDate, err := getAggregatorClientCAExpiryDate(sshRunner)
+	if err != nil {
+		return false, err
+	}
+	return time.Now().After(certExpiryDate), nil
+}
+
+func getAggregatorClientCAExpiryDate(sshRunner *ssh.Runner) (time.Time, error) {
+	certExpiryDate := time.Time{}
+	certExpiryDateCmd := `date --date="$(sudo openssl x509 -in /etc/kubernetes/static-pod-resources/kube-apiserver-certs/configmaps/aggregator-client-ca/ca-bundle.crt -noout -enddate | cut -d= -f 2)" --iso-8601=seconds`
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go:80]: Feel like we should have a helper which takes a `sshRunner` and a `path` as arguments and returns a cert expiry date (or a boolean indicating if the cert is expired).

File: pkg/crc/cluster/cluster.go [hunk 2]
```diff
@@ -291,23 +313,15 @@ func EnsurePullSecretPresentOnInstanceDisk(sshRunner *ssh.Runner, pullSecret *Pu
 	return sshRunner.CopyData([]byte(content), vmPullSecretPath, 0600)
 }
 
-func WaitforRequestHeaderClientCaFile(ocConfig oc.Config) error {
-	if err := WaitForOpenshiftResource(ocConfig, "configmaps"); err != nil {
-		return err
-	}
-
+func WaitforRequestHeaderClientCaFile(sshRunner *ssh.Runner) error {
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go:316]: `WaitFor...` while you change this.

File: pkg/crc/cluster/cluster.go [hunk 3]
```diff
@@ -291,23 +313,15 @@ func EnsurePullSecretPresentOnInstanceDisk(sshRunner *ssh.Runner, pullSecret *Pu
...
-		stdout, stderr, err := ocConfig.RunOcCommand(cmdArgs...)
+		expired, err := CheckAggregatorClientCAValidity(sshRunner)
 		if err != nil {
-			return fmt.Errorf("Failed to get request header client ca file %v: %s", err, stderr)
+			return fmt.Errorf("Failed to expiry date: %v", err)
 		}
-		if stdout == "" {
-			return &errors.RetriableError{Err: fmt.Errorf("missing .data.requestheader-client-ca-file")}
+		if expired {
+			return &errors.RetriableError{Err: fmt.Errorf("certificate still expired")}
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go:323]: Were you able to test this part of the PR? This PR is again doing 2 things at once, it checks certificate validity before doing the `waitForRequestHeaderClientCaFile`/`deletePods` dance, which is good for startup speed. But it also changes the implementation of `waitForRequestHeaderClientCaFile`, and this is not immediately obvious from the commit log. Imo we need to be careful with this part of the change as this workaround has been troublesome in the past. Maybe split it in a second commit?

Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go:323]: I tested it with an old bundle (4.5.4 from August) and the last one (4.5.14). I will split it 2 commits!

File: pkg/crc/cluster/cluster.go (hunk 1 above)
Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go:80]: Yes I'm waiting for https://github.com/code-ready/crc/pull/1532 to get this in a better shape. I opened it collect feedback on "is it a good thing or not to do?"

Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go:80]: done

File: pkg/crc/cluster/cluster.go (hunk 2 above)
Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go:316]: done

File: pkg/crc/cluster/cluster.go [hunk 4]
```diff
@@ -291,23 +313,15 @@ func EnsurePullSecretPresentOnInstanceDisk(sshRunner *ssh.Runner, pullSecret *Pu
...
+func WaitforRequestHeaderClientCaFile(sshRunner *ssh.Runner) error {
 	lookupRequestHeaderClientCa := func() error {
-		cmdArgs := []string{"get", "configmaps/extension-apiserver-authentication", `-ojsonpath={.data.requestheader-client-ca-file}`,
-			"-n", "kube-system"}
-
-		stdout, stderr, err := ocConfig.RunOcCommand(cmdArgs...)
+		expired, err := CheckAggregatorClientCAValidity(sshRunner)
 		if err != nil {
-			return fmt.Errorf("Failed to get request header client ca file %v: %s", err, stderr)
+			return fmt.Errorf("Failed to expiry date: %v", err)
```
[the remark this answers is not in the document]
Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go:320]: done

**Comment by guillaumerose** [comment]
Any thoughts on this @praveenkumar ? Should we get that in ?

File: pkg/crc/cluster/cluster.go [hunk 5]
```diff
@@ -41,16 +41,20 @@ const (
 	aggregatorClientCert = "/etc/kubernetes/static-pod-resources/kube-apiserver-certs/configmaps/aggregator-client-ca/ca-bundle.crt"
 )
 
-func CheckCertsValidity(sshRunner *ssh.Runner) (bool, bool, error) {
+func CheckCertsValidity(sshRunner *ssh.Runner) (bool, bool, bool, error) {
```
Comment (cfergeau) [inline pkg/crc/cluster/cluster.go:44]: Time for some API doc comments, or to put these 3 bools in a struct ;)

Comment (guillaumerose) [reply pkg/crc/cluster/cluster.go]: I changed it to a map, it is better! :)

**Pull request merged**